pkg/eval/testutils.go

```diff
@@ -223,6 +226,29 @@ func matchOut(want, got []interface{}) bool {
 		return false
 	}
 	for i := range got {
+		// Equality of some data types needs to be special-cased in unit tests.
```

Comment:
I like how you have moved the approximate comparison into `matchOut`, but I'd prefer to make it even more explicit by doing the following:

1. Add a new type, `type Approximately struct{ f float64 }`.

2. When `want[i]` has that type, compare with `got[i]` with an epsilon.

3. Change the unit tests that need to compare float64 approximately from:

    ```go
    That("some code").Puts(1.0)
    ```

    to

    ```go
    That("some code").Puts(Approximately{1.0})
    ```

    Most test cases will *not* need approximate comparison. You can simply run the tests, see which ones are failing, and change those to use `Approximately`.

This way, test cases that need precise will continue to do so, and test cases that need approximate comparison are very explicit.

Reply:
> Most test cases will not need approximate comparison. You can simply run the tests, see which ones are failing, and change those to use Approximately.

Yes, most test cases do not need an approximate comparison. But that is simply because the existing tests have predictable bit patterns. For example, the constant `1` has a predicable float64 bit pattern. If every time a test involving float64 values fails due to a small epsilon, and the code is changed to use `Approximately()`, what is the point?

Reply:
I'm going to go ahead and implement your suggestion but note that a test such as

```
That(`math:cos 1`).Puts(fmt.Sprintf("%g", math.Cos(1.0))),
```

still fails because the default invocation of `vals.Equal()` doesn't handle the inconsequential difference between the two values.

Reply:
> But that is simply because the existing tests have predictable bit patterns. 

It's because Elvish needs to guarantee that certain floating point number operations *have* accurate results. For example, 1.0 + 2.0 needs to be exactly 3.0 (unlike 0.1 + 0.2, which is approximately 3.0). Since Elvish does not offer an integer type yet this has to be part of what Elvish promises.

> If every time a test involving float64 values fails due to a small epsilon, and the code is changed to use Approximately(), what is the point?

I suggested that you add `Approximately` to failing tests as a quick way to fix the tests, but that's *not* the criteria for whether a test should have precise or approximate in the first place. See comment above.

> ... still fails because the default invocation of vals.Equal() doesn't handle the inconsequential difference between the two values.

The difference *is* consequential because the two values *are* different. They have different JSON presentations for example, and the tests should be precise in which form it expects.

In general, I feel you are missing an important distinction between what *an average user would expect* and *what Elvish should promise*. It's completely OK for an Elvish user to think "oh floating point numbers are messy, better to not expect anything accurate", but the same attitude is counter-productive for an Elvish developer. You need to think about all the edges cases where the subtle difference can translate into a bigger one and be as rigorous as possible.